File: code/platform/src/main/java/org/fbme/ide/platform/editor/EditorProjectionControllerRegistry.java
package org.fbme.ide.platform.editor;

import jetbrains.mps.util.containers.ConcurrentHashSet;

import java.util.Set;

public class EditorProjectionControllerRegistry {

    private static final EditorProjectionControllerRegistry INSTANCE = new EditorProjectionControllerRegistry();

    public static EditorProjectionControllerRegistry getInstance() {
        return INSTANCE;
    }

    private final Set<EditorProjectionController.Factory> myFactories = new ConcurrentHashSet<>();

    private EditorProjectionControllerRegistry() {
    }

    public void register(EditorProjectionController.Factory factory) {
        myFactories.add(factory);
    }

    public void unregister(EditorProjectionController.Factory factory) {
        myFactories.remove(factory);
    }

    public Set<EditorProjectionController.Factory> getFactories() {
        return myFactories;
    }
}

File: code/platform/src/main/kotlin/org/fbme/ide/platform/editor/EditorProjectionControllerRegistry.kt
package org.fbme.ide.platform.editor

import jetbrains.mps.util.containers.ConcurrentHashSet

class EditorProjectionControllerRegistry private constructor() {
    val factories: MutableSet<EditorProjectionController.Factory> = ConcurrentHashSet()

    fun register(factory: EditorProjectionController.Factory) {
        factories.add(factory)
    }

    fun unregister(factory: EditorProjectionController.Factory) {
        factories.remove(factory)
    }

    companion object {
        val instance = EditorProjectionControllerRegistry()
    }
}
